### ISSUE TITLE
Panic for array types in assignToRecursive when len(src.Elements) == 0

### ISSUE THREAD

**Issue by jerem-d** [jackc/pgx#881]
Hi there - 

First, thanks for the great work !

I am facing a panic while scanning rows, when at least one required column in the select query is some xxx[] array type.
My panic occurs if on one row such column is empty.

Let's look at func (src *Int8Array) AssignTo(dst interface{}) error

When execution arrives at this condition (no optimised match for my case):
```
if len(src.Elements) == 0 {
	if value.Kind() == reflect.Slice {
		value.Set(reflect.MakeSlice(value.Type(), 0, 0))
		return nil
	}
}
```

... if receiving struct is not a slice (my case, unfortunately), it continues with len(src.Elements) == 0 into this method :
`err := src.assignToRecursive(value, **0**, 0) // index=0
`
... which panics on this line, eventually:
`err := src.Elements[**index**].AssignTo(addr.Interface())
`
Would it be possible to handle such case ? Maybe not letting the execution through if len(src.Elements) == 0, just returning nil here :

```
if len(src.Elements) == 0 {
	if value.Kind() == reflect.Slice {
		value.Set(reflect.MakeSlice(value.Type(), 0, 0))
		return nil
	}
       ==> return nil
}
```

Thanks much !

**Comment by jackc** [comment]
Do you have an example that causes this failure? That would help in determining the best solution.

**Comment by jerem-d** [comment]
Something generic like this overly simplified snippet below, when in the code the query string is not known upfront.
I've played around a bit, pgx/pgxpool panic on any empty array column item as long as it skips the optimised matches and destination is not a slice (custom types, interface{}).

```
	// query := "select array_column_that_may_be_empty from my_table" // not deterministic, may vary
	rows, err := pgxPool.Query(ctx, query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	colNames := rows.Columns() // simplified
	size := len(colNames)
	vals := make([]interface{}, size)
	valptrs := make([]interface{}, size)
	for i := range vals {
		valptrs[i] = &vals[i]
	}

	for rows.Next() {
		err = rows.Scan(valptrs...) // it will panic within here on the very first empty item
		if err != nil {
			return nil, err
		}
		// handle results here
	}
```

In such case, I guess I'd expect a nil in the corresponding vals[0] - this is what some sql/database drivers on other DBs or lib/pq are returning actually.

**Comment by jackc** [comment]
Fixed in https://github.com/jackc/pgtype/commit/00d516f5c4fdecd15973cf216fb1e39716c94346.

I took the approach of returning an error instead of doing nothing in this case. Because `*interface{}` is not a valid target for an array. It already errorred if the array was not empty -- though not with a super helpful error message. I now explicitly check that the target is a slice or array at the beginning of the reflection path and error out if it is not.
